processing/src/main/java/io/druid/segment/StringDimensionIndexer.java

```diff
@@ -412,27 +415,27 @@ public IndexedInts getRow()
 
         int[] row = null;
         int rowSize = 0;
+
+        // usually due to currEntry's rowIndex is smaller than the row's rowIndex in which this dim first appears
         if (indices == null || indices.length == 0) {
           final int nullId = getEncodedValue(null, false);
           if (nullId > -1) {
-            if (nullId < maxId) {
-              row = new int[] {nullId};
-              rowSize = 1;
-            } else {
-              // Choose to use ArrayBasedIndexedInts later, instead of EmptyIndexedInts, for monomorphism
-              row = IntArrays.EMPTY_ARRAY;
-              rowSize = 0;
+            if (nullIdIntArray == null) {
+              nullIdIntArray = new int[] {nullId};
             }
+            row = nullIdIntArray;
```

Comment:
I'm good with this patch either way. It's not worse than before in terms of garbage, and it should be safe to make thread-unsafe DimensionSelectors (see also: SingleScanTimeDimSelector). @kaijianding I think it's up to you if you want to change it or not. Please let us know.